### ISSUE TITLE
Provide easy way to create an XMPP response stanza.

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2613]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)| @ralphm reported|
|-|-|
|Trac ID|trac#2613|
|Type|enhancement|
|Created|2007-05-01 12:09:47Z|

An often occurring pattern in XMPP is sending stanzas in response to a received stanza. This certainly holds for `iq` stanzas of types `get` and `set`, where you reply with an `iq` of the type `result` or `error`. The other stanzas `message` and `presence` also have an `error` type flavor to be sent in response.

<details><summary>Searchable metadata</summary>

```
trac-id__2613 2613
type__enhancement enhancement
reporter__ralphm ralphm
priority__highest highest
milestone__Words_0_6 Words-0.6
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__words words
keywords__ 
time__1178021387000000 1178021387000000
changetime__1178198290000000 1178198290000000
version__None None
owner__ralphm ralphm
cc__ralphm
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_1"></a>|@ralphm removed owner|
|-|-|

Implemented `t.w.p.j.xmlstream.toResponse`. It takes the original stanza as input, along with an optional stanza type, and creates the response stanza with given type. The response stanza has the addressing reversed and the `id` preserved.

Please review.

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_2"></a>|@ralphm commented|
|-|-|

Branch: [branches/xmpp-toresponse-2613.](branches/xmpp-toresponse-2613.)

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_3"></a>|@ralphm commented|
|-|-|

The branch also addresses [#2614](https://github.com/twisted/twisted/issues/2614), by the way.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_4"></a>|@therve set owner to @ralphm|
|-|-|

The new tests need some docstrings, and it would be cool if some could be added to modified tests.

The change in jabber.error isn't verified by a test (you probably want to check that the node returned is a new node). I also suspect that if a new node is created, the children attached should be a copy of the other children, not a direct reference.

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_5"></a>|@ralphm set owner to @therve|
|-|-|

Replying to [therve](#note_4):
> The new tests need some docstrings, and it would be cool if some could be added to modified tests.

Done.

> The change in jabber.error isn't verified by a test (you probably want to check that the node returned is a new node).

Done.

> I also suspect that if a new node is created, the children attached should be a copy of the other children, not a direct reference.

I worked on this for a while. I was tempted to do a deepcopy of the `children` list, but then remembered the presence of ticket [#7246](https://github.com/twisted/twisted/issues/7246). While looking in to that again, I noticed an issue occuring in the authentication tests for t.w.p.j.client. This occurs because there a derivative of `Element` called `IQ` is used and that stores a reference to an `XmlStream` which in turn stores a reference to the expat parser, which is not copyable.

As mentioned in [#7246](https://github.com/twisted/twisted/issues/7246), `Elements` should probably only be copied ``downwards``, but I think it goes quite out of scope to fix that in this branch. For the use cases of `StanzaError.toResponse`, which is sending out a serialized version immediately, a shallow copy of the `children` list is sufficient and I added a warning to that effect in the docstring.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_6"></a>|@therve set owner to @ralphm|
|-|-|

Looks great. The '@warning' field is not supported by pydoctor, please '@note' or just 'warning:', then merge.

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_7"></a>|@ralphm set status to `closed`|
|-|-|

(In [20178]) Provide function for creating XMPP response stanzas.

Author: ralphm
Reviewer: therve
Fixes [#2613](https://github.com/twisted/twisted/issues/2613), [#2614](https://github.com/twisted/twisted/issues/2614).

This adds `twisted.words.protocols.jabber.xmlstream.toResponse` for creating a
response stanza from a received stanza with reversed addressing. It also
ensures stanzas with partial addressing will have a proper response, which was
a problem for `twisted.words.protocols.jabber.error.StanzaError.toResponse`.
